File: src/IRC/IvcMarshaller.ts
/**
 * IvcMarshaller — mode string codec and IVC URI / response auto-unmarshaller.
 *
 * Provides the shared parsing algorithm that mirrors PHP
 * `IrcObject::parseModeStringToArray` / `arrayToModeString` / `parseSubobjects`,
 * plus HTTP response → typed IrcObject hydration.
 *
 * All methods are static — the marshaller has no instance state.
 */

import { ModeEntry } from './ModeEntry.js';
import type { IvcApiResponse, IvcParsedUri, IvcStatus, ModeMap } from './types.js';

// Forward references resolved at runtime to avoid circular imports.
// The concrete subclass constructors are registered via `IvcMarshaller.register()`.
type ObjectConstructor = { fromBody(body: IvcApiResponse): unknown };

const registry = new Map<string, ObjectConstructor>();

export class IvcMarshaller {
  // ---------------------------------------------------------------
  // Subclass registry (avoids circular imports)
  // ---------------------------------------------------------------

  /**
   * Register a prefix character → subclass mapping so `fromResponse` can
   * hydrate the correct IrcObject subtype.
   *
   *   IvcMarshaller.register('£', Network);
   *   IvcMarshaller.register('#', Channel);
   *   IvcMarshaller.register('&', Channel);
   *   IvcMarshaller.register('@', UserNick);
   */
  static register(prefix: string, ctor: ObjectConstructor): void {
    registry.set(prefix, ctor);
  }

  // ---------------------------------------------------------------
  // Mode string codec — mirrors PHP parseModeStringToArray / arrayToModeString
  // ---------------------------------------------------------------

  /**
   * Parse an IVC mode string into a `Map<key, ModeEntry | boolean>`.
   *
   * Input:  "+§maxchans=500+§motd=Welcome-t+o"
   * Output: Map {
   *   "§maxchans" → ModeEntry(plus=true,  val="500"),
   *   "§motd"     → ModeEntry(plus=true,  val="Welcome"),
   *   "t"         → ModeEntry(minus=true, val=null),
   *   "o"         → ModeEntry(plus=true,  val=null),
   * }
   *
   * Section-symbol (§) keys are always key=value pairs.
   * Single-character keys are boolean flag modes.
   * The spread operator handles multibyte Unicode correctly.
   */
  static parseModeString(raw: string): ModeMap {
    const result: ModeMap = new Map();
    const chars = [...raw.trim()]; // unicode-safe split
    let i = 0;
    let sign: '+' | '-' | '0' = '+';

    while (i < chars.length) {
      const ch = chars[i];

      if (ch === '+') { sign = '+'; i++; continue; }
      if (ch === '-') { sign = '-'; i++; continue; }
      if (ch === '0') { sign = '0'; i++; continue; }

      if (ch === '§') {
        // Consume §key[=value] until next sign delimiter
        let token = '§';
        i++;
        while (i < chars.length && chars[i] !== '+' && chars[i] !== '-' && chars[i] !== '0') {
          token += chars[i++];
        }
        const eqPos = token.indexOf('=');
        if (eqPos !== -1) {
          const key = token.substring(0, eqPos);
          const val = token.substring(eqPos + 1);
          result.set(key, new ModeEntry(sign === '+', sign === '-', val));
        } else {
          result.set(token, new ModeEntry(sign === '+', sign === '-', null));
        }
        continue;
      }

      // Single-character boolean flag mode
      result.set(ch, new ModeEntry(sign === '+', sign === '-', null));
      i++;
    }

    return result;
  }

  /**
   * Pack a ModeMap back into an IVC mode string.
   *
   * Input:  Map { "§motd" → ModeEntry(plus=true, val="Hello"), "o" → ModeEntry(plus=true) }
   * Output: "+§motd=Hello+o"
   */
  static toModeString(modes: ModeMap): string {
    let out = '';
    for (const [key, entry] of modes) {
      if (typeof entry === 'boolean') {
        out += (entry ? '+' : '-') + key;
      } else if (entry instanceof ModeEntry) {
        const sign = entry.plus ? '+' : (entry.minus ? '-' : '0');
        out += sign + key;
        if (entry.val !== null) out += '=' + entry.val;
      }
    }
    return out;
  }

  // ---------------------------------------------------------------
  // ivc:// URI parser — mirrors PHP IrcServices::parseSubobjects
  // ---------------------------------------------------------------

  /**
   * Parse an `ivc://` URI into its constituent parts.
   *
   * Examples:
   *   ivc://local.host/#fortress+ov   → { host:"local.host", prefix:"#", target:"fortress", modes:"+ov" }
   *   ivc://£§motd=Hello              → { host:"", prefix:"£", target:"", props:{ §motd:{value:"Hello"} } }
   *   ivc://:comment-1Δreactions      → { host:"", prefix:"", target:":comment-1", events:{ reactions:{} } }
   */
  static parseUri(uri: string): IvcParsedUri {
    const raw = uri;
    let rest = uri;

    // Strip scheme
    const scheme = rest.startsWith('ivc://') ? 'ivc' : '';
    if (scheme) rest = rest.slice(6);

    // Extract host (up to first /)
    let host = '';
    const slashPos = rest.indexOf('/');
    if (slashPos !== -1) {
      host = rest.slice(0, slashPos);
      rest = rest.slice(slashPos + 1);
    }

    // Object prefixes
    const PREFIXES = ['£', '$', '@', '&', '#'];
    let prefix = '';
    for (const p of PREFIXES) {
      if (rest.startsWith(p)) {
        prefix = p;
        rest = rest.slice(p.length);
        break;
      }
    }

    // Split on §prop and Δevent markers
    const SEC  = '§';
    const D1   = 'Δ';
    const D2   = '∆';

    let baseEnd = rest.length;
    for (let ci = 0; ci < [...rest].length; ci++) {
      const ch = [...rest][ci];
      if (ch === SEC || ch === D1 || ch === D2) {
        // byte offset
        baseEnd = [...rest].slice(0, ci).join('').length;
        break;
      }
    }
    let base = rest.slice(0, baseEnd);
    rest = rest.slice(baseEnd);

    // Inline modes on base target (e.g. #room+ov)
    let modes = '';
    const plusPos = base.indexOf('+');
    if (plusPos !== -1) {
      modes = base.slice(plusPos);
      base  = base.slice(0, plusPos);
    }

    // Parse §props and Δevents
    const props:  IvcParsedUri['props']  = {};
    const events: IvcParsedUri['events'] = {};

    const subChars = [...rest];
    let si = 0;
    while (si < subChars.length) {
      const sym = subChars[si];
      if (sym === SEC) {
        let token = '';
        si++;
        while (si < subChars.length && subChars[si] !== SEC && subChars[si] !== D1 && subChars[si] !== D2) {
          token += subChars[si++];
        }
        const eq = token.indexOf('=');
        const name  = eq !== -1 ? '§' + token.slice(0, eq) : '§' + token;
        const value = eq !== -1 ? token.slice(eq + 1) : '';
        props[name] = { value, modes: '' };
      } else if (sym === D1 || sym === D2) {
        let token = '';
        si++;
        while (si < subChars.length && subChars[si] !== SEC && subChars[si] !== D1 && subChars[si] !== D2) {
          token += subChars[si++];
        }
        const eq = token.indexOf('=');
        const name  = eq !== -1 ? token.slice(0, eq) : token;
        const value = eq !== -1 ? token.slice(eq + 1) : '';
        events[name] = { value, modes: '' };
      } else {
        si++;
      }
    }

    return { scheme, host, prefix, target: base, modes, props, events, raw };
  }

  // ---------------------------------------------------------------
  // Status: header parser
  // ---------------------------------------------------------------

  /**
   * Parse the IVC `Status:` response header.
   *
   * Format: "200+modes:CyberFox{subs [#room+t+o, #lobby]}"
   * Returns: { httpCode: 200, nick: "CyberFox", targets: [{ name:"#room", modes:"+t+o" }] }
   */
  static parseStatusHeader(header: string): IvcStatus {
    const raw = header.trim();
    const result: IvcStatus = { httpCode: 0, nick: '', targets: [], raw };

    const codePlusIdx = raw.indexOf('+modes:');
    if (codePlusIdx !== -1) {
      result.httpCode = parseInt(raw.slice(0, codePlusIdx), 10) || 0;
      const afterModes = raw.slice(codePlusIdx + 7); // "+modes:".length = 7

      const braceOpen  = afterModes.indexOf('{');
      const braceClose = afterModes.lastIndexOf('}');

      if (braceOpen !== -1) {
        result.nick = afterModes.slice(0, braceOpen).trim();

        if (braceClose > braceOpen) {
          const inner = afterModes.slice(braceOpen + 1, braceClose); // "subs [#room+t+o]"
          const subsMatch = inner.match(/subs\s*\[([^\]]*)\]/);
          if (subsMatch) {
            const parts = subsMatch[1].split(',').map(s => s.trim()).filter(Boolean);
            for (const part of parts) {
              const plusIdx = part.indexOf('+');
              if (plusIdx !== -1) {
                result.targets.push({ name: part.slice(0, plusIdx), modes: part.slice(plusIdx) });
              } else {
                result.targets.push({ name: part, modes: '' });
              }
            }
          }
        }
      } else {
        result.nick = afterModes.trim();
      }
    } else {
      result.httpCode = parseInt(raw, 10) || 0;
    }

    return result;
  }

  // ---------------------------------------------------------------
  // Response → IrcObject auto-unmarshal
  // ---------------------------------------------------------------

  /**
   * Hydrate the correct IrcObject subclass from an IVC API response body.
   * Returns null if no matching registered prefix is found.
   *
   * Registration happens in each subclass file via:
   *   IvcMarshaller.register('£', Network);
   */
  static fromResponse(body: IvcApiResponse): unknown | null {
    const bt = (body.base_target ?? '').trim();
    if (!bt) return null;

    const prefix = [...bt][0] ?? '';
    const ctor = registry.get(prefix);
    if (!ctor) return null;

    return ctor.fromBody(body);
  }
}

File: ivc-client/src/main/java/ivc/IvcMarshaller.java
package ivc;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * IvcMarshaller — mode string codec and IVC URI / response auto-unmarshaller.
 *
 * Provides the shared parsing algorithm that mirrors PHP
 * IrcObject::parseModeStringToArray / arrayToModeString / IrcServices::parseSubobjects,
 * plus HTTP response -> typed IrcObject hydration.
 *
 * All methods are static.
 */
public final class IvcMarshaller {

    private IvcMarshaller() {}

    // ---------------------------------------------------------------
    // Mode string codec
    // ---------------------------------------------------------------

    /**
     * Parse an IVC mode string into a Map<key, Object> where values are
     * either a {@link ModeEntry} (for §key=value pairs and single-char flags)
     * or {@code Boolean} (legacy boolean shorthand).
     *
     * Input:  "+§maxchans=500+§motd=Welcome-t+o"
     * Output: { "§maxchans" -> ModeEntry(true,false,"500"),
     *           "§motd"     -> ModeEntry(true,false,"Welcome"),
     *           "t"         -> ModeEntry(false,true,null),
     *           "o"         -> ModeEntry(true,false,null) }
     */
    public static Map<String, Object> parseModeString(String raw) {
        Map<String, Object> result = new LinkedHashMap<>();
        if (raw == null || raw.isBlank()) return result;

        // Use codepoint iteration for proper Unicode handling (§ is U+00A7).
        int[] cps = raw.codePoints().toArray();
        int i     = 0;
        char sign = '+';

        final int CP_PLUS    = '+';
        final int CP_MINUS   = '-';
        final int CP_ZERO    = '0';
        final int CP_SECTION = '\u00A7'; // §

        while (i < cps.length) {
            int cp = cps[i];

            if (cp == CP_PLUS)    { sign = '+'; i++; continue; }
            if (cp == CP_MINUS)   { sign = '-'; i++; continue; }
            if (cp == CP_ZERO)    { sign = '0'; i++; continue; }

            if (cp == CP_SECTION) {
                // Consume §key[=value] until next sign delimiter
                StringBuilder token = new StringBuilder().appendCodePoint(cp);
                i++;
                while (i < cps.length && cps[i] != CP_PLUS && cps[i] != CP_MINUS && cps[i] != CP_ZERO) {
                    token.appendCodePoint(cps[i++]);
                }
                String t    = token.toString();
                int eqPos   = t.indexOf('=');
                if (eqPos != -1) {
                    String key = t.substring(0, eqPos);
                    String val = t.substring(eqPos + 1);
                    result.put(key, new ModeEntry(sign == '+', sign == '-', val));
                } else {
                    result.put(t, new ModeEntry(sign == '+', sign == '-', null));
                }
                continue;
            }

            // Single-character boolean flag mode
            String key = new String(Character.toChars(cp));
            result.put(key, new ModeEntry(sign == '+', sign == '-', null));
            i++;
        }

        return result;
    }

    /**
     * Pack a mode map back into an IVC mode string.
     *
     * Input:  { "§motd" -> ModeEntry(true,false,"Hello"), "o" -> ModeEntry(true,false,null) }
     * Output: "+§motd=Hello+o"
     */
    public static String toModeString(Map<String, Object> modes) {
        StringBuilder sb = new StringBuilder();
        for (Map.Entry<String, Object> e : modes.entrySet()) {
            String key = e.getKey();
            Object val = e.getValue();
            if (val instanceof Boolean b) {
                sb.append(b ? '+' : '-').append(key);
            } else if (val instanceof ModeEntry me) {
                char sign = me.plus() ? '+' : (me.minus() ? '-' : '0');
                sb.append(sign).append(key);
                if (me.val() != null) sb.append('=').append(me.val());
            }
        }
        return sb.toString();
    }

    // ---------------------------------------------------------------
    // ivc:// URI parser
    // ---------------------------------------------------------------

    private static final String PREFIXES = "#&@£$";

    /**
     * Parse an {@code ivc://} URI into its constituent parts.
     */
    public static IvcParsedUri parseUri(String uri) {
        String raw    = uri;
        String rest   = uri;
        String scheme = "";

        if (rest.startsWith("ivc://")) {
            scheme = "ivc";
            rest   = rest.substring(6);
        }

        // Extract host
        String host = "";
        int slashPos = rest.indexOf('/');
        if (slashPos != -1) {
            host = rest.substring(0, slashPos);
            rest = rest.substring(slashPos + 1);
        }

        // Object prefix
        String prefix = "";
        for (char p : PREFIXES.toCharArray()) {
            if (rest.startsWith(String.valueOf(p))) {
                prefix = String.valueOf(p);
                rest = rest.substring(1);
                break;
            }
        }
        // § is multibyte: check explicitly
        if (prefix.isEmpty() && rest.startsWith("\u00A3")) { // £
            prefix = "\u00A3";
            rest = rest.substring("\u00A3".length());
        }

        // Find first subobject marker
        int secPos   = rest.indexOf('\u00A7'); // §
        int deltaPos = -1;
        int d1 = rest.indexOf('\u0394'); // Δ
        int d2 = rest.indexOf('\u221A'); // ∆ (alt)
        if (d1 != -1 && d2 != -1) deltaPos = Math.min(d1, d2);
        else if (d1 != -1) deltaPos = d1;
        else if (d2 != -1) deltaPos = d2;

        int firstSub = -1;
        if (secPos != -1 && deltaPos != -1) firstSub = Math.min(secPos, deltaPos);
        else if (secPos   != -1) firstSub = secPos;
        else if (deltaPos != -1) firstSub = deltaPos;

        String base = firstSub != -1 ? rest.substring(0, firstSub) : rest;
        String subRest = firstSub != -1 ? rest.substring(firstSub) : "";

        // Inline modes on base
        String modes = "";
        int plusPos = base.indexOf('+');
        if (plusPos != -1) {
            modes = base.substring(plusPos);
            base  = base.substring(0, plusPos);
        }

        // Props & events
        Map<String, String> props  = new LinkedHashMap<>();
        Map<String, String> events = new LinkedHashMap<>();
        parseSubobjects(subRest, props, events);

        return new IvcParsedUri(scheme, host, prefix, base, modes, props, events, raw);
    }

    private static void parseSubobjects(String s, Map<String,String> props, Map<String,String> events) {
        int[] cps = s.codePoints().toArray();
        int i = 0;
        while (i < cps.length) {
            int cp = cps[i];
            if (cp == '\u00A7') { // §
                StringBuilder token = new StringBuilder();
                i++;
                while (i < cps.length && cps[i] != '\u00A7' && cps[i] != '\u0394' && cps[i] != '\u221A') {
                    token.appendCodePoint(cps[i++]);
                }
                String t = token.toString();
                int eq = t.indexOf('=');
                String name  = eq != -1 ? "\u00A7" + t.substring(0, eq) : "\u00A7" + t;
                String value = eq != -1 ? t.substring(eq + 1) : "";
                props.put(name, value);
            } else if (cp == '\u0394' || cp == '\u221A') { // Δ / ∆
                StringBuilder token = new StringBuilder();
                i++;
                while (i < cps.length && cps[i] != '\u00A7' && cps[i] != '\u0394' && cps[i] != '\u221A') {
                    token.appendCodePoint(cps[i++]);
                }
                String t = token.toString();
                int eq = t.indexOf('=');
                String name  = eq != -1 ? t.substring(0, eq) : t;
                String value = eq != -1 ? t.substring(eq + 1) : "";
                events.put(name, value);
            } else {
                i++;
            }
        }
    }

    // ---------------------------------------------------------------
    // Status: header parser
    // ---------------------------------------------------------------

    private static final Pattern STATUS_PATTERN =
            Pattern.compile("^(\\d+)\\+modes:([^\\{]+)\\{subs\\s*\\[([^\\]]*)\\]\\}$");

    /**
     * Parse the IVC {@code Status:} response header.
     *
     * Format: "200+modes:CyberFox{subs [#room+t+o]}"
     */
    public static IvcStatus parseStatusHeader(String header) {
        if (header == null || header.isBlank()) return IvcStatus.unknown();
        String raw = header.trim();
        Matcher m  = STATUS_PATTERN.matcher(raw);
        if (!m.matches()) {
            int code = 0;
            try { code = Integer.parseInt(raw.split("\\+")[0]); } catch (NumberFormatException ignored) {}
            return new IvcStatus(code, "", List.of(), raw);
        }
        int    code    = Integer.parseInt(m.group(1));
        String nick    = m.group(2).trim();
        String subsStr = m.group(3).trim();
        List<IvcStatus.Target> targets = new ArrayList<>();
        if (!subsStr.isEmpty()) {
            for (String part : subsStr.split(",")) {
                part = part.trim();
                int plusIdx = part.indexOf('+');
                if (plusIdx != -1) {
                    targets.add(new IvcStatus.Target(part.substring(0, plusIdx), part.substring(plusIdx)));
                } else if (!part.isEmpty()) {
                    targets.add(new IvcStatus.Target(part, ""));
                }
            }
        }
        return new IvcStatus(code, nick, List.copyOf(targets), raw);
    }

    // ---------------------------------------------------------------
    // Response -> IrcObject auto-unmarshal
    // ---------------------------------------------------------------

    /**
     * Hydrate the correct IrcObject subclass from a raw parsed API response.
     * Returns null if no matching subclass is found.
     */
    public static IrcObject fromResponse(Map<String, Object> body) {
        Object bt = body.get("base_target");
        if (bt == null) return null;
        String baseTarget = bt.toString().trim();
        if (baseTarget.isEmpty()) return null;

        // Determine prefix (first codepoint)
        int first = baseTarget.codePointAt(0);
        String prefix = new String(Character.toChars(first));

        return switch (prefix) {
            case "\u00A3" -> Network.fromBody(body);   // £
            case "#", "&" -> Channel.fromBody(body);
            case "@"      -> UserNick.fromBody(body);
            default       -> null;
        };
    }
}